src/script/api/script_goal.cpp

```diff
@@ -138,7 +138,7 @@
 	EnforcePrecondition(false, ScriptGame::IsMultiplayer());
 	EnforcePrecondition(false, ScriptClient::ResolveClientID(client) != ScriptClient::CLIENT_INVALID);
 	/* Can only send 16 bits of client_id before proper fix is implemented */
-	EnforcePrecondition(false, client < (1 << 16));
+	EnforcePrecondition(false, (uint32)client < (1 << 16));
```

Comment:
Is it worth keeping this line at all? Given this can never fail even with the cast

Reply:
Based on the `CliendID` type it may be always true, but `client` can contain any value, and if there's enough join/leave a client can get a >65535 ID

Reply:
I don't see how - the `client` parameter is 16bits regardless of what it was before passed into this function. There's no way it can have a value outside of 0-65535 here. Unless of course we're planning for a time where ClientID is not 16 bits, but it's still dead code right now...

Reply:
This line was added in https://github.com/OpenTTD/OpenTTD/commit/36e4bd4023fd8fc2f73a7618708a723d11095786. Maybe we should see with @ldpl

Reply:
Well, client_id is supposed to be 32 bit. If it's not it may actually cause some issues for network server not just GS. But, thing is, it's declared as just enum so it's not really guaranteed to be 32 bit. I'm not familiar with clang so have no idea why enums are 16 bit  bit that's what should be fixed IMO, not this line. Quick googling says there is `-fno-short-enums` so I'd guess it should be used. Another option would be to use typed enums wherever it matters.

Reply:
I think there's some confusion between usage of `ScriptClient::ClientID` and the global `ClientID`, however both of them could have explicit sizes added, there's nothing stopping a compiler from picking a non-`int` size for enums. Technically speaking, you're not really supposed to use values not defined in the enum (i.e. 3+)

Simple change of `enum ClientID` -> `enum ClientID : uint32` should suffice for this, however

Reply:
@LordAro Ofc adding type is a better way but are you sure ClientID is the only enum in the game that needs 32 bits? So unless you want to check all of them I'd suggest using 32 bit enum option just in case.

Reply:
we'd have noticed if there was anything significant, I think

Reply:
this particular one is fairly new, most of the others already had their underlying type defined via TinyEnumT (which I recently removed in favour of C++11-style enum sizes)

Reply:
@LordAro it's not new, important enum here is ::ClientID one in network_type.hpp which was there for ages without anyone noticing. ScirptClient::ClientID I just copied to match it.

Reply:
yeah, i got that one as well